tiledb/sm/query/sparse_unordered_with_dups_reader.cc

```diff
@@ -564,320 +397,700 @@ Status SparseUnorderedWithDupsReader::create_result_tiles() {
+          }
+          mem_usage += tile_size;
+          rt++;
+          idx++;
+        }
+
+        {
+          std::unique_lock<std::mutex> ul(it_mtx);
+          rt_idx = std::min(idx, rt_idx);
+        }
+
+        return Status::Ok();
+      });
+  RETURN_NOT_OK_ELSE(status, logger_->status(status));
+
+  if (rt_idx == 0)
+    return Status::SparseUnorderedWithDupsReaderError(
+        "Unable to copy one tile with current budget/buffers");
+
+  *max_rt_idx = rt_idx;
+
+  return Status::Ok();
+}
+
+template <class BitmapType>
+template <class OffType>
+Status SparseUnorderedWithDupsReader<BitmapType>::copy_tiles() {
+  auto timer_se = stats_->start_timer("copy_tiles");
+
+  uint64_t max_rt_idx = 0;
+  RETURN_NOT_OK(initial_copy_bound(&max_rt_idx));
+
+  // Create the result tiles vector to use with read_tiles and unfilter_tiles.
+  std::vector<ResultTile*> tmp_result_tiles;
+  uint64_t i = 0;
+  for (auto it = result_tiles_[0].begin(); i < max_rt_idx; i++, it++) {
+    tmp_result_tiles.emplace_back(&*it);
+  }
+
+  // Copy tiles attribute by attribute.
+  uint64_t num_cells_copied = std::numeric_limits<uint64_t>::max();
+  for (auto& it : buffers_) {
+    // For easy reference.
+    const auto& name = it.first;
+    const auto is_dim = array_schema_->is_dim(name);
+    const auto var_sized = array_schema_->var_size(name);
+    const auto nullable = array_schema_->is_nullable(name);
+    const auto cell_size = array_schema_->cell_size(name);
+
+    std::vector<std::string> names = {name};
+
+    // Get dim idx for zipped coords copy.
+    auto dim_idx = 0;
+    if (is_dim) {
+      const auto& dim_names = array_schema_->dim_names();
+      while (name != dim_names[dim_idx])
+        dim_idx++;
+    }
+
+    if (!subarray_.is_set() || !is_dim) {
+      // Read and unfilter tiles.
+      RETURN_NOT_OK(read_attribute_tiles(&names, &tmp_result_tiles));
+      RETURN_NOT_OK(unfilter_tiles(name, &tmp_result_tiles));
+    }
+
+    // Compute initial cells copied.
+    if (num_cells_copied == std::numeric_limits<uint64_t>::max()) {
+      num_cells_copied = 0;
+      i = 0;
+      for (auto it = result_tiles_[0].begin(); i < max_rt_idx; i++, it++) {
+        num_cells_copied +=
+            subarray_.is_set() ? it->bitmap_num_cells : it->cell_num();
+      }
+    }
+
+    // Pointers to var size data.
+    std::vector<void*> var_data;
+    if (var_sized) {
+      var_data.resize(num_cells_copied);
+    }
+
+    // Process all tiles in parallel.
+    OffType offset_div =
+        elements_mode_ ? datatype_size(array_schema_->type(name)) : 1;
+    uint64_t global_cell_offset = 0;
+    std::mutex it_mtx;
+    auto result_tiles_it = result_tiles_[0].begin();
+    auto status = parallel_for(
+        storage_manager_->compute_tp(), 0, max_rt_idx, [&](uint64_t) {
+          // Take a result tile in the list.
+          typename std::list<ResultTileWithBitmap<BitmapType>>::iterator rt;
+          uint64_t cell_offset = 0;
+          {
```

Comment:
Needs documentation. In general, we should not use mutexes in parallel loops like this. Instead, we should calculate the work to be done beforehand and then assign to each thread in the loop without the need for locking.

Reply:
Added some documentation, let me know if it doesn't make sense.